classifier_and_summarizer/summarizer/text_media_matcher.py

```diff
@@ -0,0 +1,173 @@
+
+        self.media_description_embeddings \
+            = self.sentence_embedding_model.encode(
+                [element.element_text for element in self.media_descriptions])
+
+        self.media_attribute_embeddings \
+            = self.sentence_embedding_model.encode(
+                [element.element_text for element in self.media_attributes])
+
+    def _cosine_similarity(self, embedding_A, embedding_B):
+        '''returns the cosine similarity between two vectors'''
+        return 1.0 - cosine_distance(embedding_A, embedding_B)
+
+    def _similarity_score(self, sentence_index, media_index):
+        sentence_similarity_score \
+            = (1.0 + max(
+                self._cosine_similarity(
+                    self.summary_sentence_embeddings[sentence_index],
+                    self.media_description_embeddings[media_index]
+                ),
+                self._cosine_similarity(
+                    self.summary_sentence_embeddings[sentence_index],
+                    self.media_attribute_embeddings[media_index]
+                )
+            )
+            )
+
+        distance_score \
+            = 1.0 + abs(
+                self.media_attributes[media_index].element_index
+                - self.summary_sentences[sentence_index].element_index
+            )
+
+        return sentence_similarity_score / distance_score
+
+    def _form_preference_matrix(self):
+        ''' Builds and initializes the preference matrix for media+sentences'''
+        self._embed_text()  # find the vector embeddings for both lists
+
+        for i in range(self.set_size):
+            for j in range(self.set_size):
+                self.similarity_matrix[i][j] = self._similarity_score(
+                    sentence_index=i, media_index=j)
+
+        for i in range(self.set_size):
+            self.media_preference_for_sentence[i].sort(
+                key=lambda x: self.similarity_matrix[i][x], reverse=True)
+            self.sentence_preference_for_media[i].sort(
+                key=lambda x: self.similarity_matrix[x][i], reverse=True)
+
+    def _gale_shapley_matching(self):
+        '''
+        Finds the stable matching between the text and media
+        Given two (n,n) matrices of preferences for the set of Sentences, Media
+        finds the stable matching by running the gale-shapley
+        matching algorithm
+        Returns : a list of tuples where each tuple (x,y) means
+        x = index of sentence
+        y = index of media
+        Thus, it returns the indices matched as a list of tuples
+        '''
+        n = self.set_size
+
+        # Make the matching optimal for the Media
+        some_media_is_unmatched = True
+
+        # -1 denotes it is currently unmatched
+        sentence_matched_for_media = [-1 for i in range(n)]
+        media_matched_for_sentence = [-1 for i in range(n)]
+
+        def media_rank_in_sentence_preference(sentence_index, media_index):
```

Comment:
We can directly do self.media_preference_for_sentence[sentence_index].index(media_index) here.

Reply:
Also, nested functions could reduce readability of the code. Is there any benefit to using nested functions here?

Reply:
Thanks for the suggestion - will use the function to simplify the task,
There is no major benefit - my idea was that the function would only be used by that method and it would suffice to have local scope instead of object scope. Will update this accordingly!